pkg/kubelet/util/manager/watch_based_manager.go

```diff
@@ -177,6 +198,21 @@ func (c *objectCache) Get(namespace, name string) (runtime.Object, error) {
 		return nil, apierrors.NewNotFound(c.groupResource, name)
 	}
 	if object, ok := obj.(runtime.Object); ok {
+		// If the returned object is immutable, stop the reflector.
+		//
+		// NOTE: we may potentially not even start the reflector if the object is
+		// already immutable. However, given that:
+		// - we want to also handle the case when object is marked as immutable later
+		// - Secrets and ConfigMaps are periodically fetched by volumemanager anyway
+		// - doing that wouldn't provide visible scalability/performance gain - we
+		//   already have it from here
+		// - doing that would require significant refactoring to reflector
+		// we limit ourselves to just quickly stop the reflector here.
+		if utilfeature.DefaultFeatureGate.Enabled(features.ImmutableEphemeralVolumes) && c.isImmutable(object) {
+			if item.stop() {
```

Comment:
Is there any chance of a race between the watchHandler and this stop, such that hasSynced is set to false after the above check and so cannot be set back to true?

Reply:
hasSynced - as implemented:
```
hasSynced: func() (bool, error) { return reflector.LastSyncResourceVersion() != "", nil },
```
will never be set to false if it was already set to true.

So given that before we do anything we wait for it to become true (L189), it will never be set to false again.

As for other races - stop() is explicitly protected by its own lock, so that should work fine.